modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C

```diff
@@ -0,0 +1,635 @@
+  std::vector<const DofMap *> dof_maps(eq.n_systems());
+  for (const auto i : make_range(eq.n_systems()))
+  {
+    const auto & sys = eq.get_system(i);
+    dof_maps[i] = &sys.get_dof_map();
+  }
+
+  auto is_fi_evaluable = [this, &dof_maps](const FaceInfo & fi) {
+    if (!isFaceGeometricallyRelevant(fi))
+      return false;
+
+    auto is_elem_evaluable = [&dof_maps](const Elem & elem) {
+      for (const auto * const dof_map : dof_maps)
+        if (!dof_map->is_evaluable(elem))
+          return false;
+
+      return true;
+    };
+
+    // We definitely need the face info element to be evaluable in all cases
+    if (!is_elem_evaluable(fi.elem()))
+      return false;
+
+    if (fi.neighborPtr())
+      // We're on an internal face and interpolation to this face will just entail linear
+      // interpolation from neighboring cell values. We'll just check then if the neighbor is
+      // evaluable (we've already checked the element)
+      return is_elem_evaluable(fi.neighbor());
+
+    // Else we are a boundary face. Two-term boundary face extrapolation will require a cell value
+    // and gradient, which will require evaluations on all surrounding elements
+
+    mooseAssert(fi.elem().on_boundary(),
+                "If we don't have a neighbor pointer, then this element has to be on a boundary");
+
+    for (auto * const neighbor : fi.elem().neighbor_ptr_range())
+    {
+      if (!neighbor)
+        continue;
+
+      if (!is_elem_evaluable(*neighbor))
+        return false;
+    }
+
+    return true;
+  };
+
+  for (const auto & fi : all_fi)
+    if (is_fi_evaluable(fi))
+      _evaluable_fi.push_back(&fi);
+
+  _evaluable_fi.shrink_to_fit();
+}
+
+void
+INSFVRhieChowInterpolator::residualSetup()
+{
+  if (!_initial_setup_done)
+    insfvSetup();
+
+  _initial_setup_done = true;
+}
+
+void
+INSFVRhieChowInterpolator::meshChanged()
+{
+  insfvSetup();
+}
+
+void
+INSFVRhieChowInterpolator::initialize()
+{
+  _elements_to_push_pull.clear();
+
+  _a.clear();
+  _b.clear();
+  _b2.clear();
+}
+
+void
+INSFVRhieChowInterpolator::execute()
+{
+  // A lot of RC data gathering leverages the automatic differentiation system, e.g. for linear
+  // operators we pull out the 'a' coefficents by querying the ADReal residual derivatives member at
+  // the element or neighbor dof locations. Consequently we need to enable derivative computation.
+  // We do this here outside the threaded regions
+  const auto saved_do_derivatives = ADReal::do_derivatives;
+  ADReal::do_derivatives = true;
+
+  PARALLEL_TRY
+  {
+    GatherRCDataElementThread et(_fe_problem, _var_numbers);
+    Threads::parallel_reduce(*_elem_range, et);
+  }
+  PARALLEL_CATCH;
+
+  PARALLEL_TRY
+  {
+    using FVRange = StoredRange<std::vector<const FaceInfo *>::const_iterator, const FaceInfo *>;
+    GatherRCDataFaceThread<FVRange> fvr(_fe_problem, _var_numbers);
+    FVRange faces(_fe_problem.mesh().faceInfo().begin(), _fe_problem.mesh().faceInfo().end());
```

Comment:
This is guaranteed to be a subset of the `evaluable_fi`. This will only be local faces, e.g. it will not contain "evaluable" ghost faces that are included in `evaluable_fi`. This ensures that we are not calling duplicate computation of our residual objects for a given face.